src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage.MessageInterception;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Export(typeof(MessageInterceptor))]
+    [InterceptMethod(Methods.TextDocumentPublishDiagnosticsName)]
+    internal class RazorHtmlPublishDiagnosticsInterceptor : MessageInterceptor
+    {
+        private readonly LSPRequestInvoker _requestInvoker;
+        private readonly LSPDocumentManager _documentManager;
+        private readonly LSPDocumentMappingProvider _documentMappingProvider;
+        private readonly LSPDocumentSynchronizer _documentSynchronizer;
+
+        [ImportingConstructor]
+        public RazorHtmlPublishDiagnosticsInterceptor(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentManager documentManager,
+            LSPDocumentMappingProvider documentMappingProvider,
+            LSPDocumentSynchronizer documentSynchronizer)
+        {
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            if (documentSynchronizer is null)
+            {
+                throw new ArgumentNullException(nameof(documentSynchronizer));
+            }
+
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _documentMappingProvider = documentMappingProvider;
+            _documentSynchronizer = documentSynchronizer;
+        }
+
+        public override async Task<InterceptionResult> ApplyChangesAsync(JToken token, string containedLanguageName, CancellationToken cancellationToken)
+        {
+            if (token is null)
+            {
+                throw new ArgumentNullException(nameof(token));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var diagnosticParams = token.ToObject<VSPublishDiagnosticParams>();
+
+            if (diagnosticParams is null)
+            {
+                throw new ArgumentException("Conversion of token failed.");
+            }
+
+            // We only support interception of Virtual HTML Files
+            if (!RazorLSPConventions.IsVirtualHtmlFile(diagnosticParams.Uri))
```

Comment:
@jimmylewis given we can just do this is the "containedLanguageName" actually valuable in the `ApplyChangesAsync`?

Reply:
Just double checked this, it is used in their message interception layer for mapping purposes. Seeing as we use the same API I think this'll have to be kept in place.

Reply:
Yep, it is unfortunately necessary based on how we do the document mapping, but it's also informative for you as to the originating context.  For example, `foo.razor__virtual.html__virtual.css` (with a range from css-a to css-b), will get first transformed by the HTML interceptor to look like a diagnostic from `foo.razor__virtual.html` (with a range from html-a to html-b).  Then you guys can make the determination to morph it again to `foo.razor` (from razor-a to razor-b), or do any other specific changes you want, while also knowing that this originated from a CSS (or whatever other language opts in to middle layer interception).